### ISSUE TITLE
Representations: Allow "public" vs. "admin" URL generation

### ISSUE THREAD

**Issue by zerocrates** [omeka/omeka-s#343]
The representations' URL generation currently is set up to produce API URLs and admin-side URLs. This won't be suitable for use on the public-side pages, or for linking from the admin to the public, or vice-versa.

The easiest route here is probably to just create a third URL method that's just for making public URLs. Alternatives include having some automatic code and/or flag for the existing url method that tells whether public or admin URLs should be generated (and having the default be whatever side is current).

**Comment by jimsafley** [comment]
Items are currently the only resources that need URL generation on the public side since item is the only resource with a site show page. Can you think of other resources that should have their own show page? Perhaps media?

**Comment by zerocrates** [comment]
It may be necessary to be able to have sets generate a public link: to the public browse page restricted to results from that set.

As for media, there's a distinct possibility/probability that we do indeed need a page for them.

Then, of course, there's the more obvious resources that also need public links: Sites and Pages.

**Comment by jimsafley** [comment]
Here's one solution that requires only one change to `AbstractResourceRepresentation::url()`: append `$public = null` flag to the signature. Representations that have a public url will override the method, check against the current route and the flag, and either hand off to `parent::url()` to generate an admin URL or generate a public URL. Something like this in `ItemRepresentation`:

```php
public function url($action = null, $canonical = false, $public = null)
{
    $routeMatch = $this->getServiceLocator()->get('Application')
        ->getMvcEvent()->getRouteMatch();
    if ($routeMatch->getParam('__ADMIN__') && !$public) {
        return parent::url($action, $canonical);
    }

    $url = $this->getViewHelper('Url');
    return $url(
        'site/show',
        array(
            'site-slug' => $action,
            'id' => $this->id(),
        ),
        array('force_canonical' => $canonical)
    );
}
``` 

This assumes that admin URLs will not be generated while on a public route, which may be a wrong assumption.

**Comment by zerocrates** [comment]
A problem occurs to me: an Item link can't be merely "public," it has to be in reference to a specific Site. (Whereas the "public" link for a Site or Page can just be public, they internally have all the information they need to generate that link.) I see you're trying to stuff the site-slug into the `$action`, which I'm not totally comfortable with (ditto for the inclusion of the `$public` param that will be ignored by the base implementation). That pattern would also make the public-side Item URL call uniquely clunky because you'd have to pass it the site slug whenever you wanted to use it.

Some sort of "automatic" url generation that detects whether you're on the admin or some specific site and does the URL correctly to stay in that context should be okay, but it may be cleanest to totally separate that from the underlying functions for generating admin and/or public URLs, so you could always just directly use one of those if you needed to cross the boundary between public and admin or vice-versa.

**Comment by jimsafley** [comment]
Fixed in 50cd256e15da1b571166aebab389a861deaea416
